File: reuse/C++/Algorithms/Sort/quick_sort.cpp
/* =============================================================================
> File Name: quick_sort.cpp
> Author: DaiDai
> Mail: daidai4269@aliyun.com
> Created Time: Wed 04 Dec 2019 09:43:20 PM CST
============================================================================= */


#include <iostream>
#include <vector>

using namespace std;


void quick_sort_(vector<int>& arr, int start, int end) {
	if (start >= end)
		return ;
	int temp = arr[start];
	int left = start + 1, right = end;
	while (left < right) {
		while (arr[left] < temp && left < right)
			left++;
		while (arr[right] >= temp && left < right)
			right--;
		swap(arr[left], arr[right]);
	}
	if (arr[right] <= temp)
		swap(arr[right], arr[start]);
	else
		right++;
	quick_sort_(arr, start, left - 1);
	quick_sort_(arr, left, end);
}

void quick_sort(vector<int> &A) {
	quick_sort_(A, 0, A.size() - 1);
}

int main() {
	vector<int> vec = {5, 3124, 2345, 56, 67, 45, 34};
	quick_sort(vec);
	for (const int& n : vec)
		cout << n << " ";
	cout << endl;
}

File: reuse/python3/algorithm/sort/quick_sort.py
#!/usr/bin/env python3
# -*- coding:utf-8 -*-


# =============================================================================
# File Name: quick_sort.py
# Author: DaiDai
# Mail: daidai4269@aliyun.com
# Created Time: Mon 28 Oct 2019 09:20:52 PM CST
# =============================================================================


def quick_sort(lst):
    def sort(lst, first, last):
        if first >= last:
            return 
        mid_val = lst[first]
        low = first
        high = last
        while low < high:
            while low < high and lst[high] >= mid_val:
                high -= 1
            lst[low] = lst[high]
            while low < high and lst[low] < mid_val:
                low += 1
            lst[high] = lst[low]
        lst[low] = mid_val
        sort(lst, first, low - 1)
        sort(lst, low + 1, last)
        return lst

    return sort(lst, 0, len(lst) - 1)

if __name__ == "__main__":
    lst = [9, 8, 7, 6, 5, 4, 3, 2, 1, 0]
    l = quick_sort(lst)
    print(l)

